expose scopes to perm_xxx_replicas permissions

jouvin opened the pull request:
Fixes #8333.

Originally linked to #8288 closed in favor of #8333.

jouvin committed:
expose scopes to perm_xxx_replicas permissions

- add/delete_replicas() + update_replicas_states() refactoring to use a
  common function for checking permissions, to ensure consistency
- Add the scope list to the parameters passed to the permission functions

fix previous mod

jouvin commented:
It is my attempt to fix the problem that scopes are not passed to the `has_permission()` function when adding (or deleting) a replica when they are supposed to be the Rucio object on which permissions are applied when uploading or downloading DIDs. 

It is more intrusive than #8236 as it is necessary to build the scope list in `gateway/replica.py` and pass it in the `kwargs` argument to the permission module (`core/permission/generic.py` for example). The alternative would be to require the permission module to look in all replicas passed to get the scope and to what is done once in `gateway/replica.py` so IMO it is less optimal. And anyway scopes are not passed at all in `update_replicas_states()` in `gateway/replica.py` so a modification of this module would be necessary anyway...

I am not enough a Rucio expert to ensure that there is no side effect to this modification but it's my feeling: nothing passed previously is modified, just an entry `scopes` is added in the `kwargs` dict so it should be harmless. And the resulting default permission of this PR is a basic configuration that you can use to test your Rucio installation, including upload operations, without writing a permission policy as your first step when configuring a new instance. And production instance probably all have their own policy file so they will not be impacted by the updated `generic.py` (a similar update is required in `generic_muti_vo.py` once we agreed on `generic.py`). I think that exposing scopes to replica-related permission functions could also benefit to some production instance that had to work around the absence of scopes with account attributes....

Interested by feedback from experts!

maxnoe commented on lib/rucio/core/permission/generic.py:
```diff
@@ -689,20 +689,29 @@ def perm_declare_suspicious_file_replicas(issuer: "InternalAccount", kwargs: dic
-    return str(kwargs.get('rse', '')).endswith('SCRATCHDISK')\
-        or str(kwargs.get('rse', '')).endswith('USERDISK')\
+
+    if str(kwargs.get('rse', '')).endswith('SCRATCHDISK')\
```
Maybe refactor this a bit with early exits instead of ors to make it easier to reason about?

And maybe a constant for the `SPECIAL_RSE_SUFFIXES` 

```python
rse = kwargs.get("rse", "")

if _is_root(issuer):
    return True
if has_account_attribute(account=issuer, key='admin', session=session):
   return True
if any(rse.endswith(suffix) for suffix in SPECIAL_RSE_SUFFIXES):
    return True

# Also allow if user owns all affected scopes
if all(is_scope_owner(scope, issuer, session=session) for scope in kwargs["scopes"]):
    return True

return False
```

jouvin replied:
I agree with your proposal but not that the special RSE suffixes are already there and it is not my addition...

bziemons commented:
As part of an effort to keep the number of pull requests under control, your PR will be closed due to:

* The CI checks are failing and there has been no activity for over seven days.
* A review requested changes and there has been no activity addressing these for over seven days.

Please be advised that this is not a definitive rejection of your effort. Feel free to reopen it once the aforementioned reason has been addressed.

jouvin commented:
@bziemons sorry, I have been away from Rucio since the beginning of the year but I hope to restart in the coming weeks and will try to fix the problems.

bziemons commented:
> @bziemons sorry, I have been away from Rucio since the beginning of the year but I hope to restart in the coming weeks and will try to fix the problems.

no need to be sorry :) We just need to get on top of PRs again, so closing the ones that haven't moved for a while helps us focusing our reviewing efforts and will get PRs merged more quickly. Please re-open when the PR is ready to review as mentioned in the last sentence of my closing comment

jouvin commented:
@maxnoe I pushed changes addressing your remarks and suggestions. How I can reopen the PR?

Not sure why the new commit pushed doesn't show up... May be because the PR is closed...

I also don't know what I should do about Codecov report.

maxnoe commented:
You should have a button here, right?

jouvin commented:
No I don't... May be only admins can do it...

jouvin commented:
@bziemons I'd like to reopen this PR but I don't find how to do it. I don't have the button to reopen the PR...

bziemons commented:
> @bziemons I'd like to reopen this PR but I don't find how to do it. I don't have the button to reopen the PR...

Unfortunately, me neither:

<img width="492" height="81" alt="image" src="https://github.com/user-attachments/assets/9747c40f-7acc-41fe-a64e-8a7e0ac85a75" />

I have never seen that message. @bari12 can you re-open the PR?

jouvin commented:
Yes, I force-pushed the source branch. I never had a problem with this but may be GitHub changed something when it happens and the PR is closed... I found this post that seems somewhat related: https://gist.github.com/robertpainsi/2c42c15f1ce6dab03a0675348edd4e2c

May be the easiest is to forget this PR... I'll create a new one and reference this one but I don't think there is so much information useful here. Mainly @maxnoe suggestions that I implemented.

bari12 commented:
Same for me, it shows the same message and no re-open button. I can only imagine something happened with the origin branch which makes GH think that this will never be mergable?

jouvin commented:
I'll  open a new PR as suggested earlier

dchristidis commented:
To reopen a PR, the branch must be in the exact state it was at the time it was closed. If you’ve already modified, you can push the original branch: `git push --force origin d094f08:scope_in_upload_generic_permission` (unless the housekeeping of the local repository has already cleaned it).
